include/shared/game/servergamestate.hpp

```diff
@@ -0,0 +1,280 @@
+	 */
+	unsigned int getTimestep() const;
+
+	/**
+	 * @brief Returns the timestep length of this ServerGameState instance.
+	 * @return Timestep length (in milliseconds)
+	 */
+	std::chrono::milliseconds getTimestepLength() const;
+
+	/**
+	 * @brief Returns the phase that this ServerGameState instance is currently
+	 * in.
+	 * @return The current GamePhase of this ServerGameState instance. 
+	 */
+	GamePhase getPhase() const;
+
+	/**
+	 * Reassign id to the specified name in the mapping. This is okay to call if the
+	 * player is already in the mapping, as nothing will happen. If a player's name
+	 * has changed, then this will update their name as well.
+	 */
+	void addPlayerToLobby(EntityID id, std::string name);
+	/**
+	 * Removes a player from the lobby with the specified id.
+	 */
+	void removePlayerFromLobby(EntityID id);
+	/**
+	 * Getter for the mapping between entity ID and player name in the lobby
+	 */
+	const std::unordered_map<EntityID, std::string>& getLobbyPlayers() const;
+	/**
+	 * Returns how many max players can be in the lobby, based on the config option
+	 */
+	int getLobbyMaxPlayers() const;
+
+	/*	Debugger Methods	*/
+
+	/**
+	 * @brief Creates a string representation of this ServerGameState object.
+	 * @return A string representation of this ServerGameState object.
+	 */
+	std::string to_string();
+
+private:
+	/*
+	 * Note on how Objects are stored:
+	 *
+	 * The ServerGameState class has a vector<Object &> objects array that holds
+	 * references to all Objects in the game instance at the current timestep.
+	 * The actual objects are stored in type-specific vectors, e.g., all of the
+	 * Player objects are stored in vector<Player> players, and all Wall objects
+	 * are stored in vector<Wall> walls; this idea requires that each object is
+	 * stored in exactly one type-specific object vector. I think a good rule is
+	 * that an object should be stored in the type-specific array whose type is
+	 * closest to its actual type. E.g., let's say Creature derives from Object
+	 * and Player and Enemy derive from Creature, and that there exists a
+	 * vector<Object> base_objects type-specific object vector as well as a 
+	 * vector<Creature> creatures type-specific object vector. All Player, 
+	 * Enemy, and Creature objects should be stored in the creatures vector and 
+	 * all other objects should be stored in the base_objects vector.
+	 * 
+	 * To access an object in the vector<Object &> vector, use its global ID as
+	 * the index.
+	 * To access an object in the type-specific vector that contains it, use
+	 * its type-specific ID as the index.
+	 * Hence, all Objects have two IDs - a global ID and a type-specific ID.
+	 */
+
+	//	TODO: Refactor vector<T> array + queue<unsigned int> freelist into a
+	//	single template class - otherwise, each type-specific object vector will
+	//	come with an extra free list and this will get hairy to use.
+
+	SmartVector<Object *> objects;
```

Comment:
I like this approach but I wanted to clarify the need for the `SmartVector`class. Is it to hide the implementation details of managing the freelist in `SmartVector`

Reply:
Yes, exactly - currently, each object has two IDs, a global ID and a type-specific ID; the idea here is that the `ServerGameState` class would have vectors for each object type (e.g., `Player`, `Creature`, etc.) and objects of that type would be stored in these vectors. The type-specific ID of the object would be its index in this type-specific vector, and its global ID would be its index in a global `vector<Object &> objects` vector. To avoid wasting space, when objects get removed and then new objects get created, I'd like for the freed object IDs to get reassigned. This means that every vector will need a free list to keep track of available IDs, and since we may have as many of these vectors as there will be object types, I think it makes sense to encapsulate the vectors with their free lists in a new class, which here is `SmartVector`.